tools/clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -5097,14 +5097,65 @@ static void ProcessDeclAttribute(Sema &S, Scope *scope, Decl *D,
   }
 }
 
+void Sema::ValidateShaderAttribtues(Scope *S, Decl *D, const AttributeList *A) {
+  bool hasComputeAttr = false;
+  bool hasNodeAttr = false;
+  int attrNum = 0;
+
+  for (const AttributeList *l = A; l; l = l->getNext()) {
```

Comment:
This pattern is also not great. Rather than validating this after we've constructed an attribute list, we can do it during construction.

If you look at how `handleWorkGroupSize` is implemented. Basically it checks the decl to see if an attribute specifying the work group has already been attached, if it has then it verifies the work group size matches the new specification.

In this case you could see if there is already an attribute, and if there is verify that it is the `node`/`compute` combination that is valid, otherwise issue a diagnostic.